amaranth/hdl/_ir.py

```diff
@@ -372,208 +349,137 @@ def _propagate_domains(self, missing_domain, *, platform=None):
-                    for value in [port._addr, port._en, port._data]:
-                        subfrag.add_ports(value._rhs_signals(), dir="i")
-                        add_uses(value._rhs_signals())
-                for domain, _ in subfrag.iter_sync():
-                    cd = subfrag.domains[domain]
-                    add_uses(cd.clk)
-                    if cd.rst is not None:
-                        add_uses(cd.rst)
-            else:
-                parent[subfrag] = self
-                level [subfrag] = level[self] + 1
-
-                subfrag._prepare_use_def_graph(parent, level, uses, defs, ios, top)
-
-    def _propagate_ports(self, ports, all_undef_as_ports):
-        # Take this fragment graph:
-        #
-        #    __ B (def: q, use: p r)
-        #   /
-        #  A (def: p, use: q r)
-        #   \
-        #    \_ C (def: r, use: p q)
-        #
-        # We need to consider three cases.
-        #   1. Signal p requires an input port in B;
-        #   2. Signal r requires an output port in C;
-        #   3. Signal r requires an output port in C and an input port in B.
-        #
-        # Adding these ports can be in general done in three steps for each signal:
-        #   1. Find the least common ancestor of all uses and defs.
-        #   2. Going upwards from the single def, add output ports.
-        #   3. Going upwards from all uses, add input ports.
-
-        parent = {self: None}
-        level  = {self: 0}
-        uses   = _ast.SignalDict()
-        defs   = _ast.SignalDict()
-        ios    = _ast.SignalDict()
-        self._prepare_use_def_graph(parent, level, uses, defs, ios, self)
-
-        ports = _ast.SignalSet(ports)
-        if all_undef_as_ports:
-            for sig in uses:
-                if sig in defs:
-                    continue
-                ports.add(sig)
-        for sig in ports:
-            if sig not in uses:
-                uses[sig] = set()
-            uses[sig].add(self)
-
-        @memoize
-        def lca_of(fragu, fragv):
-            # Normalize fragu to be deeper than fragv.
-            if level[fragu] < level[fragv]:
-                fragu, fragv = fragv, fragu
-            # Find ancestor of fragu on the same level as fragv.
-            for _ in range(level[fragu] - level[fragv]):
-                fragu = parent[fragu]
-            # If fragv was the ancestor of fragv, we're done.
-            if fragu == fragv:
-                return fragu
-            # Otherwise, they are at the same level but in different branches. Step both fragu
-            # and fragv until we find the common ancestor.
-            while parent[fragu] != parent[fragv]:
-                fragu = parent[fragu]
-                fragv = parent[fragv]
-            return parent[fragu]
-
-        for sig in uses:
-            if sig in defs:
-                lca  = reduce(lca_of, uses[sig], defs[sig])
+        ports = self._fixup_ports(ports)
```

Comment:
I think this doesn't need to be a method--it's just called once, here, and should probably be inlined.

Reply:
(I think the other `_prepare_x` methods are methods because I wanted to test them separately from the rest of the implementation.)